### DISCUSSION TITLE
Using SES Modules in browser environment

### DISCUSSION THREAD

**Discussion by vamshich13** [endojs/endo#1550] [General]
Hello,

Can the @endo/static-module-record package be used in a browser environment, or is it only intended for use in a Node.js environment? 
I am getting this error when I use @endo/static-module-record in browser environment:
```
ReferenceError: process is not defined
    at eval (core.ts?9295:40:17)
    at ../common/temp/node_modules/.pnpm/@babel+types@7.19.0/node_modules/@babel/types/lib/definitions/core.js 
```
If it is only intended for use in Node.js, then how can I use SES modules in a browser environment?

Please reply to my query, [ @naugtur @kriskowal @michaelfig ]
Thanks

**Comment by naugtur** [comment]
Hello,

SES and most tools built on top of it are runtime agnostic. @endo/static-module-record  should work in the browser.

What is `core.ts` and how is the error above related to `@endo/static-module-record` ?
It doesn't look like the error is coming from @endo/static-module-record.

**Reply by vamshich13** [reply]
Thanks for the response. @naugtur 
Looks like the `@endo/static-module-record ` is using this package `@babel/types`, which uses `process` in
`@babel/types/lib/definitions/core.js
` which is designed for use in Node.js

**Reply by naugtur** [reply]
That's not what the stacktrace says. Can you share more context? Are you using a bundler? How do you use @endo/static-module-record? Is your work in a private repository somewhere? Or can you create a minimal reproduction repo?

**Reply by vamshich13** [reply]
Hello @naugtur  
I have created minimal reproduction repo: [sample_module](https://github.com/vamshich13/sample_module)
Please check

**Comment by michaelfig** [comment]
@vamshich13 I see that you must be bundling since the `@babel/` NPM packages use CommonJS `require`. They also reference `process`, but for the dependencies of `@endo/static-module-record`, they only seem to use `process.env[someEnvironmentVariable]`.  You can neutralize these references by doing:

```js
globalThis.process = { env: {} };
```

somewhere in an `import`ed module before you import `@endo/static-module-record`.

**Reply by vamshich13** [reply]
@michaelfig 
Thanks for the response, I have tried this.
But I got another error:

```
Uncaught ReferenceError: Buffer is not defined
    at eval (jsesc.js:30:18)
    at ./node_modules/@agoric/babel-generator/node_modules/jsesc/jsesc.js (bundle.js:228:1)
    at __webpack_require__ (bundle.js:2329:42)
    at eval (types.js:27:14)
    at ./node_modules/@agoric/babel-generator/lib/generators/types.js (bundle.js:140:1)
    at __webpack_require__ (bundle.js:2329:42)
    at eval (index.js:72:14)
    at ./node_modules/@agoric/babel-generator/lib/generators/index.js (bundle.js:74:1)
    at __webpack_require__ (bundle.js:2329:42)
    at eval (printer.js:10:26)
```

**Reply by naugtur** [reply]
I was wrong, the only reason why it works for us in the browser is how we bundle it.

agoric/babel-generator specifies  "jsesc": "^2.5.1". jsesc declares only versions prior to 2 are browser compatible. 
Looks like we're bundling polyfills for some basics when we're using static-module-record.
Browserify used to polyfill these automatically. If you're using webpack, you probably need to set fallbacks
 https://viglucci.io/articles/how-to-polyfill-buffer-with-webpack-5
This would address process too.

**Reply by naugtur** [reply]
This commit https://github.com/mathiasbynens/jsesc/commit/3c0362b2a794d06bdb99358e2d369fb4ebb71ebd says we should update to latest jsesc in @agoric/babel-generator

**Reply by vamshich13** [reply]
Thanks So much for the help! 
It worked with webpack fallback.

```

  resolve: {
    fallback: {
        process: require.resolve('process/browser'),
        buffer: require.resolve('buffer/'),
    },
  },
...
  plugins: [
    new HtmlWebpackPlugin(),
    new webpack.ProvidePlugin(
      {
        process: 'process/browser',
        Buffer: ['buffer', 'Buffer'],
      }
    ),
  ],
```

**Reply by vamshich13** [reply]
@naugtur 
I have one more question. 
https://github.com/endojs/endo/discussions/1576
